### ISSUE TITLE
Download numbers have different meaning for libraries and apps

### ISSUE THREAD

**Issue by kornelski** [rust-lang/crates.io#1481]
I've been looking at using download numbers for crate ranking, and found that it strongly favors libraries over applications. Top apps have an order of magnitude smaller download numbers. 

I suspect that's because use of `cargo install` is rare, apps have other distribution channels, aren't built in CI that often, aren't redownloaded after `cargo update`, etc.

I wonder if something could be done about this, e.g. 

 1. Could you count `cargo install` download numbers separately? (just for the installed crate, not its deps. It doesn't even have to be broken down per version, since usual semver locking problems don't apply to app installs). 
    I can't reliably detect from crate data whether the crate is an app. There are exceptions like crates that are both app and a library. The `cc` crate has a binary, but nobody uses it this way.
   With this data I could estimate a fudge factor by which install numbers have to be multiplied to be comparable with library download numbers.

 2. Could `cargo` detect when it's in CI (e.g. by looking for [env vars](https://github.com/watson/ci-info/blob/70691ac0e08085896273593002398e3ce439f7fd/index.js#L5-L30)) and download dependencies in a way that isn't counted? I suspect that would eliminate one large source of library downloads inflation.

 3. Could library downloads be counted only when a library is downloaded for the first time after being added to `Cargo.toml`? (when it wasn't in a lockfile?)

**Comment by jtgeibel** [comment]
Some of these may be possible, but I'm not sure how it would interact with the caching of crates that is already done by cargo.  For example, in case 1, if I `cargo install` a crate that uses `libc` and then start a new project that uses the same version of `libc`, then the download wouldn't be counted at all since cargo has already cached the download.

I agree that CI is probably the biggest cause of inflated download counts, for several reasons.  First, at least on Travis, if you test on all 3 channels these are treated independently as far as caching is concerned.  So if you add a dependency in a PR, that dependency will be downloaded 3 times.  In my experience, the cache also isn't 100% reliable on Travis, so pushing new commits may cause dependencies to be downloaded in duplicate again.  Finally, PRs and branches have separate caches so dependencies will be downloaded yet again when the PR is merged to master.

On the crates.io code base, we test on all 3 channels plus a fixed nightly for clippy.  We also use `bors-ng` for merging PRs, which tests the merge on a staging branch and then again on master.  So bumping a version or adding a crate causes it to be downloaded 12 times in the best case, in addition to downloads on developer machines and the Heroku deploy.

In summary, I think option 2 would be the most straightforward option and have the largest impact.

**Comment by sgrif** [comment]
Download count is always going to be an inherently flawed number. It is easily gamed, and inflated by a variety of sources. This issue mentions that we should try to filter out CI, but why? The number is describing exactly what it says -- the number of times a crate has been downloaded. I'm not sure why some downloads should be counted, but others should not.

It seems like download count is being conflated with user count here. They are not the same thing. We do not have an accurate way to measure user count, nor should we pretend that we are ever going to be able to do that. I certainly don't think we should be performing what some would consider a surprising invasion of privacy to arbitrarily filter downloads that we think "shouldn't count".

Either way, since this metric is used for one of the sorting options on crates.io, and also to determine what crates are included on play.rust-lang.org, I think any major change to how we measure that would require a formal RFC.

My personal opinion is that we should not pretend that download count is something that it isn't, and any attempt to do that just makes a flawed metric both more flawed, and more dishonest about what's being measured.

**Comment by kornelski** [comment]
Download numbers are used on crates.io and play.rust-lang.org, because [it's *already* assumed it's a proxy for number of users of a crate](https://github.com/rust-lang/rfcs/blob/master/text/1824-crates.io-default-ranking.md): 

> downloads show how many people or other crates have found this crate to be worthy of using.

In both of these places the goal was to choose "popular" crates, and not absolutely literally crates with most HTTP requests from whatever source. So while the "number of downloads" is describing exactly what it is, it's not used, and it is not interpreted as exactly what it is. This discrepancy is a problem.

If your objection is only that it's inappropriate to call "download numbers" a number that isn't exactly counting every download, then I'm proposing to rename it or add another metric for under another name, let's say "popularity".

**Comment by kornelski** [comment]
But you use that number for ranking on crates.io. That is a bad ranking signal!

**Comment by sgrif** [comment]
It is not a ranking, it is a metric that we describe very concretely. Either way, I do not think we are going to change each other's minds. I'm going to step away from this conversation.

**Comment by Turbo87** [comment]
I'll close this issue here for now since there hasn't been much activity lately and it seems unlikely to change in the near future.
